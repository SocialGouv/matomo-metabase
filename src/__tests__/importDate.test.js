const { importDate } = require("../importDate");

const mock_pgQuery = jest.fn();
const mock_matomoApi = jest.fn();

process.env.MATOMO_SITE = "42";
process.env.PROJECT_NAME = "some-project";

const matomoVisit = require("./visit.json");

jest.mock("pg", () => {
  class Client {
    escapeIdentifier(name) {
      return name;
    }
    end() {}
    connect() {
      return Promise.resolve();
    }
    query(...args) {
      return mock_pgQuery(...args);
    }
  }
  return {
    Client,
  };
});

jest.mock("piwik-client", () => {
  class MatomoClient {
    api(...args) {
      return mock_matomoApi(...args);
    }
  }
  return MatomoClient;
});

beforeEach(() => {
  jest.resetAllMocks();
  jest.resetModules();
  process.env.STARTDATE = "";
});

test("importDate: should import all events for a given date", async () => {
  const pgSpy = jest.fn().mockResolvedValue();
  const matomoSpy = jest.fn().mockImplementation((options, cb) => {
    return cb(null, [
      {
        ...matomoVisit,
        idVisit: 123,
      },
      {
        ...matomoVisit,
        idVisit: 124,
      },
    ]);
  });
  const fakeClient = {
    escapeIdentifier(name) {
      return name;
    },
    query: pgSpy,
  };
  await importDate(fakeClient, matomoSpy, new Date("2021-08-02"));
  expect(matomoSpy.mock.calls.map((c) => c[0])).toMatchSnapshot();
  expect(pgSpy.mock.calls).toMatchSnapshot();
});

test("importDate: should import when new results", async () => {
  const pgResult = { rows: [] };
  const pgSpy = jest.fn().mockResolvedValue(pgResult);
  const matomoSpy = jest.fn().mockImplementation((options, cb) => {
    return cb(null, [
      {
        ...matomoVisit,
        idVisit: 123,
      },
    ]);
  });
  const fakeClient = {
    query: pgSpy,
    escapeIdentifier(name) {
      return name;
    },
  };

  const imported = await importDate(fakeClient, matomoSpy, new Date("2021-08-03T00:00:00"));
  expect(pgSpy.mock.calls.length).toEqual(3);
  expect(imported.length).toEqual(2);
  expect(matomoSpy.mock.calls[0][0].filter_offset).toEqual(0);
});

test("importDate: should offset matomo calls when results already exist", async () => {
  const pgResult = { rows: [{ count: 42 }] };
  const pgSpy = jest.fn().mockResolvedValue(pgResult);
  const matomoSpy = jest.fn().mockImplementation((options, cb) => {
    return cb(null, [
      {
        ...matomoVisit,
        idVisit: 123,
      },
    ]);
  });
  const fakeClient = {
    query: pgSpy,
    escapeIdentifier(name) {
      return name;
    },
  };

  const imported = await importDate(fakeClient, matomoSpy, new Date("2021-08-03T00:00:00"));
  expect(pgSpy.mock.calls.length).toEqual(3);
  expect(imported.length).toEqual(2);
  expect(matomoSpy.mock.calls[0][0].filter_offset).toEqual(42);
});

test("importDate: should NOT import when no matomo results", async () => {
  const pgSpy = jest.fn().mockResolvedValue({ rows: [{ action_timestamp: 1627948800000 }] });
  const matomoSpy = jest.fn().mockImplementation((options, cb) => {
    return cb(null, []);
  });
  const fakeClient = {
    query: pgSpy,
    escapeIdentifier(name) {
      return name;
    },
  };
  const imported = await importDate(fakeClient, matomoSpy, new Date("2021-08-03"));
  expect(pgSpy.mock.calls.length).toEqual(1);
  expect(imported.length).toEqual(0);
});
