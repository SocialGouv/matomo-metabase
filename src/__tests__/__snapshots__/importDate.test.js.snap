// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`importDate: should import all events for a given date 1`] = `
Array [
  Object {
    "date": "2021-08-02",
    "filter_limit": 500,
    "filter_offset": 0,
    "filter_sort_order": "asc",
    "idSite": 0,
    "method": "Live.getLastVisitsDetails",
    "period": "day",
  },
]
`;

exports[`importDate: should import all events for a given date 2`] = `
Array [
  Array [
    "SELECT COUNT(distinct idvisit) FROM matomo WHERE action_timestamp::date='2021-08-02';",
  ],
  Array [
    "insert into matomo
        (idsite, idvisit, actions, country, region, city, operatingsystemname, devicemodel, devicebrand, visitduration, dayssincefirstvisit, visitortype, sitename, userid, serverdateprettyfirstaction, action_id, action_type, action_title, action_eventcategory, action_eventaction, action_eventname, action_eventvalue, action_timespent, action_timestamp, action_url, siteSearchKeyword, usercustomproperties, usercustomdimensions)
        values ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20, $21, $22, $23, $24, $25, $26, $27, $28)
        ON CONFLICT DO NOTHING",
    Array [
      "42",
      123,
      "2",
      "Argentine",
      "Buenos Aires",
      "Buenos Aires",
      "Mac",
      "Générique Bureau",
      "Inconnu",
      "300",
      "23",
      "returningCustomer",
      "tests",
      "24",
      "2021-08-20T21:55:12.000Z",
      "123_0",
      "event",
      "Evènement",
      "Ecommerce",
      "Cart change",
      "added - Basic Wetsuit",
      1,
      48,
      "2021-08-20T21:35:18.000Z",
      "https://dive-shop.net/products/basic-wetsuit/",
      undefined,
      Object {
        "page-author": "Julien",
        "post-age": "-430 days",
      },
      Object {
        "dimension1": "guest",
        "dimension2": "julien",
        "dimension3": "page",
        "dimension4": "indonesia",
        "dimension5": "diving",
      },
    ],
  ],
  Array [
    "insert into matomo
        (idsite, idvisit, actions, country, region, city, operatingsystemname, devicemodel, devicebrand, visitduration, dayssincefirstvisit, visitortype, sitename, userid, serverdateprettyfirstaction, action_id, action_type, action_title, action_eventcategory, action_eventaction, action_eventname, action_eventvalue, action_timespent, action_timestamp, action_url, siteSearchKeyword, usercustomproperties, usercustomdimensions)
        values ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20, $21, $22, $23, $24, $25, $26, $27, $28)
        ON CONFLICT DO NOTHING",
    Array [
      "42",
      123,
      "2",
      "Argentine",
      "Buenos Aires",
      "Buenos Aires",
      "Mac",
      "Générique Bureau",
      "Inconnu",
      "300",
      "23",
      "returningCustomer",
      "tests",
      "24",
      "2021-08-20T21:55:12.000Z",
      "123_1",
      "action",
      "Divezone Brand Diving Boots - Divezone Store",
      undefined,
      undefined,
      undefined,
      undefined,
      "2",
      "2021-08-20T21:30:25.000Z",
      "https://dive-shop.net/products/diving-boots/",
      undefined,
      Object {},
      Object {
        "dimension1": "guest",
        "dimension2": "julien",
        "dimension3": "page",
        "dimension4": "indonesia",
        "dimension5": "diving",
      },
    ],
  ],
  Array [
    "insert into matomo
        (idsite, idvisit, actions, country, region, city, operatingsystemname, devicemodel, devicebrand, visitduration, dayssincefirstvisit, visitortype, sitename, userid, serverdateprettyfirstaction, action_id, action_type, action_title, action_eventcategory, action_eventaction, action_eventname, action_eventvalue, action_timespent, action_timestamp, action_url, siteSearchKeyword, usercustomproperties, usercustomdimensions)
        values ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20, $21, $22, $23, $24, $25, $26, $27, $28)
        ON CONFLICT DO NOTHING",
    Array [
      "42",
      123,
      "2",
      "Argentine",
      "Buenos Aires",
      "Buenos Aires",
      "Mac",
      "Générique Bureau",
      "Inconnu",
      "300",
      "23",
      "returningCustomer",
      "tests",
      "24",
      "2021-08-20T21:55:12.000Z",
      "123_2",
      "search",
      "Divezone Brand Diving Boots - Divezone Store",
      undefined,
      undefined,
      undefined,
      undefined,
      "2",
      "2021-08-20T21:30:22.000Z",
      "https://dive-shop.net/products/diving-boots/",
      "scuba",
      Object {},
      Object {
        "dimension1": "guest",
        "dimension2": "julien",
        "dimension3": "page",
        "dimension4": "indonesia",
        "dimension5": "diving",
      },
    ],
  ],
  Array [
    "insert into matomo
        (idsite, idvisit, actions, country, region, city, operatingsystemname, devicemodel, devicebrand, visitduration, dayssincefirstvisit, visitortype, sitename, userid, serverdateprettyfirstaction, action_id, action_type, action_title, action_eventcategory, action_eventaction, action_eventname, action_eventvalue, action_timespent, action_timestamp, action_url, siteSearchKeyword, usercustomproperties, usercustomdimensions)
        values ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20, $21, $22, $23, $24, $25, $26, $27, $28)
        ON CONFLICT DO NOTHING",
    Array [
      "42",
      124,
      "2",
      "Argentine",
      "Buenos Aires",
      "Buenos Aires",
      "Mac",
      "Générique Bureau",
      "Inconnu",
      "300",
      "23",
      "returningCustomer",
      "tests",
      "24",
      "2021-08-20T21:55:12.000Z",
      "124_0",
      "event",
      "Evènement",
      "Ecommerce",
      "Cart change",
      "added - Basic Wetsuit",
      1,
      48,
      "2021-08-20T21:35:18.000Z",
      "https://dive-shop.net/products/basic-wetsuit/",
      undefined,
      Object {
        "page-author": "Julien",
        "post-age": "-430 days",
      },
      Object {
        "dimension1": "guest",
        "dimension2": "julien",
        "dimension3": "page",
        "dimension4": "indonesia",
        "dimension5": "diving",
      },
    ],
  ],
  Array [
    "insert into matomo
        (idsite, idvisit, actions, country, region, city, operatingsystemname, devicemodel, devicebrand, visitduration, dayssincefirstvisit, visitortype, sitename, userid, serverdateprettyfirstaction, action_id, action_type, action_title, action_eventcategory, action_eventaction, action_eventname, action_eventvalue, action_timespent, action_timestamp, action_url, siteSearchKeyword, usercustomproperties, usercustomdimensions)
        values ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20, $21, $22, $23, $24, $25, $26, $27, $28)
        ON CONFLICT DO NOTHING",
    Array [
      "42",
      124,
      "2",
      "Argentine",
      "Buenos Aires",
      "Buenos Aires",
      "Mac",
      "Générique Bureau",
      "Inconnu",
      "300",
      "23",
      "returningCustomer",
      "tests",
      "24",
      "2021-08-20T21:55:12.000Z",
      "124_1",
      "action",
      "Divezone Brand Diving Boots - Divezone Store",
      undefined,
      undefined,
      undefined,
      undefined,
      "2",
      "2021-08-20T21:30:25.000Z",
      "https://dive-shop.net/products/diving-boots/",
      undefined,
      Object {},
      Object {
        "dimension1": "guest",
        "dimension2": "julien",
        "dimension3": "page",
        "dimension4": "indonesia",
        "dimension5": "diving",
      },
    ],
  ],
  Array [
    "insert into matomo
        (idsite, idvisit, actions, country, region, city, operatingsystemname, devicemodel, devicebrand, visitduration, dayssincefirstvisit, visitortype, sitename, userid, serverdateprettyfirstaction, action_id, action_type, action_title, action_eventcategory, action_eventaction, action_eventname, action_eventvalue, action_timespent, action_timestamp, action_url, siteSearchKeyword, usercustomproperties, usercustomdimensions)
        values ($1, $2, $3, $4, $5, $6, $7, $8, $9, $10, $11, $12, $13, $14, $15, $16, $17, $18, $19, $20, $21, $22, $23, $24, $25, $26, $27, $28)
        ON CONFLICT DO NOTHING",
    Array [
      "42",
      124,
      "2",
      "Argentine",
      "Buenos Aires",
      "Buenos Aires",
      "Mac",
      "Générique Bureau",
      "Inconnu",
      "300",
      "23",
      "returningCustomer",
      "tests",
      "24",
      "2021-08-20T21:55:12.000Z",
      "124_2",
      "search",
      "Divezone Brand Diving Boots - Divezone Store",
      undefined,
      undefined,
      undefined,
      undefined,
      "2",
      "2021-08-20T21:30:22.000Z",
      "https://dive-shop.net/products/diving-boots/",
      "scuba",
      Object {},
      Object {
        "dimension1": "guest",
        "dimension2": "julien",
        "dimension3": "page",
        "dimension4": "indonesia",
        "dimension5": "diving",
      },
    ],
  ],
]
`;
